hazelcast/src/main/java/com/hazelcast/internal/nio/Packet.java

```diff
@@ -96,6 +98,16 @@
      */
     public static final int FLAG_JET_FLOW_CONTROL = 1 << 1;
 
+    /**
```

Comment:
Why do we need 2 bits? One bit would be sufficient.

Reply:
We have these special members both on 3.x and 4.x which can process _some_ packets both from 3.x and 4.x members. To do that, they rely on these flags. Let's say this special 4.x member says "if the FLAG_4_0 is absent, it's sent by 3.x and if it's present, it's sent by 4.x". This means, all members in the cluster that it's part of need to set this flag for the cluster to operate normally and this allows 3.8+ clusters to connect to this special member.

Now, we have the other side of the replication, for users wanting to replicate back from 4.x to 3.x. They need to run a special 3.12 member. To distinguish incoming packets from 3.12.x members and 4.x, we need to again check some flag. Let's say we don't have the FLAG_3_x and let's say, as you proposed we check "if the FLAG_4_0 is present, it's sent by 4.x and if it's absent, it's sent by 3.x". This places a restriction on the 4.x cluster -> it has to now be comprised out of members setting this flag, namely 4.0.2+.

So, to sum up, it's a decision on what kind of restriction we want to impose on users:
a) if we use just one flag like FLAG_4_0, the restriction would be "your 4.x cluster, be it the source or target of migration, has to be at least 4.0.2+"
b) if we use use two flags, the restriction would be "your target cluster which contains the special migration member has to be at least 4.0.2+ or 3.12.8+, depending on whether you're replicating 3.x->4.x or 4.x->3.x)

Also, I don't see using two flags as a big loss. Sure, it introduces a cognitive load but that's it. The bits are already there, they are just not used. Also, since migration tool will probably not support migrating to 4.2, we'll remove all traces of these flags in 4.2.

Reply:
In even shorter terms, the option with one flag places a restriction on the versions of both clusters while the option with two flags places a restriction on the versions of only one cluster.